addons/font/xml.c

```diff
@@ -0,0 +1,138 @@
+#include "allegro5/allegro.h"
+#include <ctype.h>
+
+#include "xml.h"
+
+typedef struct {
+   XmlState state;
+   bool closing;
+   ALLEGRO_USTR *value;
+   int (*callback)(XmlState state, char const *value, void *u);
+   void *u;
+} XmlParser;
+
+static void scalar(XmlParser *x) {
+   x->callback(x->state, al_cstr(x->value), x->u);
+   al_ustr_assign_cstr(x->value, "");
+}
+
+static void opt_scalar(XmlParser *x) {
+   if (al_ustr_size(x->value)) {
+      scalar(x);
+   }
+}
+
+static void discard_scalar(XmlParser *x) {
+   al_ustr_assign_cstr(x->value, "");
+}
+
+static void close_tag(XmlParser *x) {
+   x->state = Outside;
+   x->closing = false;
+}
+
+static void add_char(XmlParser *x, char c) {
+   char s[] = {c, 0};
+   al_ustr_append_cstr(x->value, s);
+}
+
+static void create_tag(XmlParser *x) {
+   scalar(x);
+}
+
+static void open_tag(XmlParser *x) {
+   x->state = Outside;
+}
+
+void _al_xml_parse(ALLEGRO_FILE *f,
```

Comment:
Is this a full XML parser or some subset of the language?

Reply:
subset